subscription-fee/src/fees.rs

```diff
@@ -0,0 +1,142 @@
+use auto_farm::common::{
+    address_to_id_mapper::{AddressId, AddressToIdMapper},
+    unique_payments::UniquePayments,
+};
+
+multiversx_sc::imports!();
+
+#[multiversx_sc::module]
+pub trait FeesModule {
+    #[only_owner]
+    #[endpoint(addAcceptedFeesTokens)]
+    fn add_accepted_fees_tokens(
+        &self,
+        accepted_tokens: MultiValueEncoded<EgldOrEsdtTokenIdentifier>,
+    ) {
+        for token in accepted_tokens {
+            require!(token.is_valid(), "Invalid token");
+
+            self.accepted_fees_tokens().insert(token);
+        }
+    }
+
+    #[payable("*")]
+    #[endpoint(depositFees)]
+    fn deposit_fees(&self) {
+        let (payment_token, payment_amount) = self.call_value().egld_or_single_fungible_esdt();
+        require!(payment_amount > 0, "No payment");
+        require!(
+            self.accepted_fees_tokens().contains(&payment_token),
+            "Invalid payment token"
+        );
+
+        let caller = self.blockchain().get_caller();
+        let caller_id = self.user_id().get_id_or_insert(&caller);
+        self.add_user_payment(
+            caller_id,
+            EgldOrEsdtTokenPayment::new(payment_token, 0, payment_amount),
+            self.user_deposited_fees(caller_id),
+        );
+    }
+
+    #[endpoint(withdrawFunds)]
+    fn withdraw_funds(
+        &self,
+        tokens_to_withdraw: MultiValueEncoded<MultiValue2<EgldOrEsdtTokenIdentifier, BigUint>>,
+    ) -> MultiValue2<BigUint, ManagedVec<EsdtTokenPayment>> {
+        let caller = self.blockchain().get_caller();
+        let caller_id = self.user_id().get_id_non_zero(&caller);
+
+        let user_fees_mapper = self.user_deposited_fees(caller_id);
+        let mut all_user_tokens = user_fees_mapper.get().into_payments();
+        let mut egld_amount = BigUint::zero();
+        let mut output_payments = ManagedVec::new();
+        for pair in tokens_to_withdraw {
+            let (token_id, amount) = pair.into_tuple();
+
+            if token_id.is_egld() {
+                let user_egld_amount = self.user_deposited_egld(caller_id).take();
+                if user_egld_amount > 0 {
+                    self.send().direct_egld(&caller, &user_egld_amount);
+                    egld_amount = user_egld_amount;
+                }
+
+                continue;
+            }
+
+            let mut opt_found_token_index = None;
+            for (index, user_token) in all_user_tokens.iter().enumerate() {
+                if user_token.token_identifier == token_id && user_token.amount >= amount {
+                    output_payments.push(user_token);
+                    opt_found_token_index = Some(index);
+                    break;
+                }
+            }
+
+            if opt_found_token_index.is_none() {
+                continue;
+            }
+
+            let token_index = unsafe { opt_found_token_index.unwrap_unchecked() };
```

Comment:
Just an idea (can be totally wrong)

```
if Some(token_index) = opt_found_token_index{
...
}
```

I get it that continue looks more clean but isn't it more scary to have an `unsafe` inside a contract endpoint

Reply:
No, that leads to way too much nested code. Also, it's not unsafe, but the Rust compiler is not (yet) smart enough to realize that.